File: 066_SubsetsII.cpp
// Topic: Subsets II
// Description: Generate unique subsets avoiding duplicates.
// Approach: Sort input and backtrack skipping duplicates at the same recursion level.

#include <bits/stdc++.h>
using namespace std;

class Solution
{
   public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums)
    {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        vector<int> cur;
        function<void(int)> dfs = [&](int idx)
        {
            res.push_back(cur);
            for (int i = idx; i < (int)nums.size(); ++i)
            {
                if (i > idx && nums[i] == nums[i - 1])
                    continue;
                cur.push_back(nums[i]);
                dfs(i + 1);
                cur.pop_back();
            }
        };
        dfs(0);
        return res;
    }
};

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    vector<int> a = {1, 2, 2};
    auto res = Solution().subsetsWithDup(a);
    for (auto& v : res)
    {
        for (int i = 0; i < (int)v.size(); ++i)
        {
            if (i)
                cout << " ";
            cout << v[i];
        }
        cout << "\n";
    }
}

/*
Time Complexity: O(n * 2^n) worst-case (output size)
Space Complexity: O(n) recursion depth + output
*/

File: 066_SubsetsII.py
# Topic: Subsets II
# Description: Generate unique subsets avoiding duplicates.
# Approach: Sort input and backtrack skipping duplicates at the same recursion level.

from typing import List


class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        result = []
        cur = []

        def dfs(idx):
            result.append(cur[:])
            for i in range(idx, len(nums)):
                if i > idx and nums[i] == nums[i - 1]:
                    continue
                cur.append(nums[i])
                dfs(i + 1)
                cur.pop()

        dfs(0)
        return result


if __name__ == "__main__":
    nums = [1, 2, 2]
    result = Solution().subsetsWithDup(nums)
    for v in result:
        print(v)

"""
Time Complexity: O(n * 2^n) worst-case (output size)
Space Complexity: O(n) recursion depth + output
"""
